I0078DO-164 BON specification migration (#945)
MicroEJ/docs: ApplicationDeveloperGuide/bon.rst

If the application code does not change, the order remains the same.
The order is also compatible with the Java semantic. Intuitively,
a class may depend on other classes. Those classes should be initialized
first. We list a few of these dependencies: object creation, superclass,
methods receiver, arguments and fields types, ….

Dependencies of classes upon themselves define a graph of dependencies.
This graph may depict cycles. The graph is linearized in an order which
depends only on the graph itself.

The classes dependencies also includes all the classes of pre-configured
:ref:`immutable objects <immutable>`  objects.

Utilities
---------

Timer & TimerTask
~~~~~~~~~~~~~~~~~

An `Timer`_ defines a single Java thread in charge of scheduling
`Runnable`_ objects from the `TimerTask`_ class. All `TimerTask`_
are executed sequentially, according to their schedule. A `Timer`_ does
its best effort to schedule the `TimerTask`_ appropriately, which depends
on the `TimerTask`_ durations and schedules (there is no real-time
guaranties).

A `TimerTask`_ may be scheduled repeatedly. In that case, the delay for

> the next schedule may depends on the end of the previous ending of the

Reviewer: ```suggestion
the next schedule may depend on the end of the previous ending of the
```

are not impacted: the `TimerTask`_ is assumed to have terminated its
execution regularly, and is not rescheduled event if it was scheduled
repeatedly.

The main APIs are:

-  `Timer.schedule(TimerTask, long)`_ and `Timer.schedule(TimerTask, Date)`_ methods
   allow to schedule one execution after the specified delay.
-  `Timer.schedule(TimerTask task, long, long)`_ and `Timer.schedule(TimerTask, Date,
   long)`_ methods allow to schedule repeatedly executions, the first
   one after the specified delay. The waiting time between two
   executions is relative to the end of the previous execution.
-  `Timer.scheduleAtFixedRate(TimerTask task, long, long)`_ and
   `Timer.scheduleAtFixedRate(TimerTask, Date, long)`_ methods allow to
   schedule repeatedly executions, the first one after the specified
   delay. The waiting time between two executions is independent of
   the end of the previous execution.

Platform Time
~~~~~~~~~~~~~

The application time is the user time: it depends on its localization.
`System.currentTimeMillis()`_ returns the application time
expressed in milliseconds since midnight, January 1, 1970 UTC.

> BON introduces a platform time that is independent from any user
> considerations: it materializes the running time since the very last
> start of the device. This time cannot be changed.

Reviewer: ```suggestion
BON introduces a platform time that is monotonic. 
It always moves forward and is not impacted by application time modifications 
(e.g., NTP or Daylight Savings Time updates).
It can be implemented by returning the running time since the start of the device.
```

Reviewer 2: ```suggestion
BON introduces a platform time that is monotonic. 
It always moves forward and is not impacted by application time modifications 
(e.g., Synchronization of time with an NTP server, TimeZone or Daylight Savings Time updates).
It can be implemented by returning the running time since the start of the device.
```

application time and platform time: 

-  `Util.platformTimeNanos()`_ and `Util.platformTimeMillis()`_ method return the
   platform time, a ``long``, expressed in nanoseconds and in
   milliseconds.
-  `Util.setCurrentTimeMillis(long)`_ and `Util.setCurrentTimeMillis(Date)`_
   methods allow to change the application time in order to match a
   user localization. This has no effect on the platform time.
   `Util.currentTimeMillis()`_ method is a synonym of
   `System.currentTimeMillis()`_.

Byte Array Accesses
~~~~~~~~~~~~~~~~~~~

The addresses space is 8-bit oriented even if there are platforms that
manipulate quantities that are larger than an 8-bit: 32-bit processors
for example do so. The ordering of individual addressable sub-components
within the representation of a larger data item is called the
endianness. ``BigEndian`` describes an ordering with the most significant
byte first, whereas ``LittleEndian`` describes an ordering with the least
significant byte first.

.. figure:: images/bon_spec/endianness.png
  :align: center
  :scale: 75%

  Representation of the 32-bit quantity 0x0000100A using both BigEndian and in LittleEndian layout.

> BON introduces methods to read and write into array of byte (byte[])

Reviewer: ```suggestion
BON introduces methods to read and write into array of byte (``byte[]``)
```

according to the platform endianness, or according to a specific
provided endianness. The `ByteArray`_ class provides such APIs:

-  `ByteArray.getPlatformEndianness()`_ returns the underlying system-dependent
   endianness, which mostly depends on the target processor(s).
-  `ByteArray.readInt(byte[], int)`_ and `ByteArray.writeInt(byte[], int, int)`_ reads and
   writes an ``int`` using the platform specific endianness.
-  `ByteArray.readInt(byte[], int, int)`_ and `ByteArray.writeInt(byte[], int, int, int)`_
   reads and writes an ``int`` using the specified endianness as last
   argument, which may be either `LITTLE_ENDIAN`_ or `BIG_ENDIAN`_.

> Similar methods are provided for ``short``, ``char``, ``long`` types.

Reviewer: ```suggestion
Similar methods are provided for ``short``, ``char``, and ``long`` types.
```
